akka-http/src/main/java/akka/http/javadsl/server/HttpApp.java

```diff
@@ -0,0 +1,155 @@
+   */
+  public void startServer(String host, int port, ServerSettings settings) throws ExecutionException, InterruptedException {
+    startServer(host, port, settings, Optional.empty());
+  }
+
+  /**
+   * Start a server on the specified host and port, using provided settings and [[ActorSystem]].
+   * Note that this method is blocking.
+   */
+  public void startServer(String host, int port, ServerSettings settings, ActorSystem system) throws ExecutionException, InterruptedException {
+    startServer(host, port, settings, Optional.of(system));
+  }
+
+  /**
+   * Start a server on the specified host and port, using provided settings and [[ActorSystem]] if present.
+   * Note that this method is blocking.
+   * This method may throw an {@link ExecutionException} or {@link InterruptedException} if the future that signals that
+   * the server should shutdown is interrupted or cancelled.
+   */
+  public void startServer(String host, int port, ServerSettings settings, Optional<ActorSystem> system) throws ExecutionException, InterruptedException {
+
+    final ActorSystem theSystem = system.orElseGet(() -> ActorSystem.create(Logging.simpleName(this).replaceAll("\\$", "")));
+    systemReference.set(theSystem);
+    final ActorMaterializer materializer = ActorMaterializer.create(theSystem);
+
+    CompletionStage<ServerBinding> bindingFuture = Http
+      .get(theSystem)
+      .bindAndHandle(route().flow(theSystem, materializer),
+        ConnectHttp.toHost(host, port),
+        settings,
+        theSystem.log(),
+        materializer);
+
+    bindingFuture.handle((binding, exception) -> {
+      if (exception != null) {
+        postHttpBindingFailure(exception);
+      } else {
+        //setting the server binding for possible future uses in the client
+        serverBinding.set(binding);
+        postHttpBinding(binding);
+      }
+      return null;
+    });
+
+    try {
+      bindingFuture
+        .thenCompose(ignore -> waitForShutdownSignal(theSystem)) // chaining both futures to fail fast
+        .toCompletableFuture()
+        .exceptionally(ignored -> Done.getInstance()) // If the future fails, we want to complete normally
+        .get(); // It's waiting forever because maybe there is never a shutdown signal
+    } finally {
+      bindingFuture.thenCompose(ServerBinding::unbind).handle( (success, exception) -> {
+        postServerShutdown(Optional.ofNullable(exception), theSystem);
+        if (!system.isPresent()) {
+          // we created the system. we should cleanup!
+          theSystem.terminate();
+        }
+        return null;
+      });
+    }
+
+  }
+
+  /**
+   * It tries to retrieve the {@link ServerBinding} if the server has been successfully started. It throws an {@link IllegalStateException} otherwise.
+   * You can use this method to attempt to retrieve the {@link ServerBinding} at any point in time to, for example, stop the server due to unexpected circumstances.
+   */
+  ServerBinding binding() {
```

Comment:
Ideally I would return a Try as well, but as there is none in the Std. lib, I was doubting was was the closest, if throwing or Optional.
I picked throwing because this way the caller of this method doesn't lose any information, like which was the exception thrown. If it would return `Optional<ServerBinding>` this information would be lost.